### DISCUSSION TITLE
How is Rune language executed in low level?

### DISCUSSION THREAD

**Discussion by evan0greenup** [rune-rs/rune#577] [Q&A]
Is Rune language firstly interpreted by Interpreter? Or firstly translate into bytecode, then executed in vm? Or compile to machine code, and directly run by machine?

**Comment by udoprog** [comment]
The language is compiled and translated to bytecode (or an array of instructions) which are executed sequentially. Memory is stored on a stack. The core logic of the virtual machine is actually [fairly easy to follow here](https://github.com/rune-rs/rune/blob/main/crates/rune/src/runtime/vm.rs#L2941).

You can see the instructions by checking "Instructions" [in the playground](https://rune-rs.github.io/play/?c=Zm4gZmliKG4pIHsKICAgIGlmIG4gPD0gMSB7CiAgICAgICAgbgogICAgfSBlbHNlIHsKICAgICAgICBmaWIobiAtIDEpICsgZmliKG4gLSAyKQogICAgfQp9CgpwdWIgZm4gbWFpbigpIHsKICAgIGZpYigxNSkKfQo%3D).

**Reply by evan0greenup** [reply]
@udoprog , how is the performance is it slower than Rust? If so, how much slower?

**Reply by udoprog** [reply]
You should expect an order of magnitude slower than any language executing natively, and about an order of magnitude faster than AST walkers / eval loops.

A little bit faster than Python and a little bit slower than Lua in my experience.
